In [ ]:
ANOVA NOT APPLICABLE AS BOTH INPUT VARIABLES HAVE ONLY 2 CLASSES (LungCap Dataset)

In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols #ordinary least square

In [2]:
#Three Hypothesis in two way Anova 
#H0: Population Means of Mean of Smoke(Y) = Population Mean of Smoke(N) 
#H1: Population Means of Mean of Gender (M) = Population of Mean of Gender (F) 
#H2: There is no interaction between two factors ie Smoke & Gender are Independent 
#ie two set of factors are Independent

In [3]:
df=pd.read_excel('Lung_Capacity.xlsx')
df

Age  Height Smoke  Gender  LungCap
0      6    62.1    no    male    6.475
1     18    74.7   yes  female   10.125
2     16    69.7    no  female    9.550
3     14    71.0    no    male   11.125
4      5    56.9    no    male    4.800
..   ...     ...   ...     ...      ...
720    9    56.0    no  female    5.725
721   18    72.0   yes    male    9.050
722   11    60.5   yes  female    3.850
723   15    64.9    no  female    9.825
724   10    67.7    no    male    7.100

[725 rows x 5 columns]

In [4]:
#perform one-way ANOVA
model = ols('LungCap ~ C(Smoke)',data=df).fit()
sm.stats.anova_lm(model, typ=2)

sum_sq     df         F    PR(>F)
C(Smoke)    52.723971    1.0  7.507152  0.006297
Residual  5077.748480  723.0       NaN       NaN

In [5]:
p_value = 0.006297

In [6]:
if p_value < 0.05:
    print('Reject Null Hypotesis')
else:
    print('Fail to reject Null Hypothesis')

Reject Null Hypotesis


In [7]:
#perform two-way ANOVA
model = ols('LungCap ~ C(Smoke) + C(Gender) + C(Smoke):C(Gender)', data=df).fit()
sm.stats.anova_lm(model, typ=2)

sum_sq     df          F    PR(>F)
C(Smoke)              62.783450    1.0   9.202814  0.002503
C(Gender)            158.021638    1.0  23.162851  0.000002
C(Smoke):C(Gender)     0.919587    1.0   0.134793  0.713621
Residual            4918.807254  721.0        NaN       NaN

In [8]:
#For Smoke
p_value = 0.002503

In [18]:
if p_value < 0.05:
    print('Reject Null Hypotesis')
else:
    print('Fail to reject Null Hypothesis')

Reject Null Hypotesis


In [19]:
#For Gender
p_value=0.000002
if p_value < 0.05:
    print('Reject Null Hypotesis')
else:
    print('Fail to reject Null Hypothesis')


Reject Null Hypotesis


In [20]:
#For Interaction term
p_value=0.713621
if p_value < 0.05:
    print('Reject Null Hypotesis')
else:
    print('Fail to reject Null Hypothesis')


Fail to reject Null Hypothesis


Post HOC test
However, this doesn’t tell us which groups are different from each other. It simply tells us that not all of the group means are equal. In order to find out exactly which groups are different from each other, we must conduct a post hoc test.

One of the most commonly used post hoc tests is Tukey’s Test, which allows us to make pairwise comparisons between the means of each group while controlling for the family-wise error rate.

In [2]:
import pandas as pd
import numpy as np
from scipy.stats import f_oneway
from statsmodels.stats.multicomp import pairwise_tukeyhsd

In [7]:
#enter data for three groups
a = [85, 86, 88, 75, 78, 94, 98, 79, 71, 80]
b = [91, 92, 93, 90, 97, 94, 82, 88, 95, 96]
c = [79, 78, 88, 94, 92, 85, 83, 85, 82, 81]


In [3]:
#perform one-way ANOVA
f_oneway(a, b, c)

F_onewayResult(statistic=5.167774552944481, pvalue=0.012582197136592609)

We can see that the overall p-value from the ANOVA table is 0.01258.

Since this is less than .05, we have sufficient evidence to say that the mean values across each group are not equal.

Thus, we can proceed to perform Tukey’s Test to determine exactly which group means are different.

To perform Tukey’s test in Python, we can use the pairwise_tukeyhsd() function from the statsmodels library:

In [4]:
#create DataFrame to hold data
df = pd.DataFrame({'score': [85, 86, 88, 75, 78, 94, 98, 79, 71, 80,
                             91, 92, 93, 90, 97, 94, 82, 88, 95, 96,
                             79, 78, 88, 94, 92, 85, 83, 85, 82, 81],
                   'group': np.repeat(['a', 'b', 'c'], repeats=10)})
df

score group
0      85     a
1      86     a
2      88     a
3      75     a
4      78     a
5      94     a
6      98     a
7      79     a
8      71     a
9      80     a
10     91     b
11     92     b
12     93     b
13     90     b
14     97     b
15     94     b
16     82     b
17     88     b
18     95     b
19     96     b
20     79     c
21     78     c
22     88     c
23     94     c
24     92     c
25     85     c
26     83     c
27     85     c
28     82     c
29     81     c

In [5]:
# perform Tukey's test
tukey = pairwise_tukeyhsd(endog=df['score'],
                          groups=df['group'],
                          alpha=0.05)

In [6]:
#display results
print(tukey)

 Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower    upper  reject
-----------------------------------------------------
     a      b      8.4 0.0158   1.4272 15.3728   True
     a      c      1.3 0.8864  -5.6728  8.2728  False
     b      c     -7.1 0.0453 -14.0728 -0.1272   True
-----------------------------------------------------


Here’s how to interpret the output:

P-value for the difference in means between a and b: .0158
P-value for the difference in means between a and c: .8864
P-value for the difference in means between b and c: .0453
Thus, we would conclude that there is a statistically significant difference between the means of groups a and b and groups b and c, but not a statistically significant difference between the means of groups a and c.